src/lib/string_pool.rs

```diff
@@ -0,0 +1,509 @@
+    /// self.0 is never `None`
+    pub(crate) fn clear(&mut self) {
+        let mut inner_pool = self.0.take();
+
+        let mut bump = inner_pool.unwrap().into_head();
+
+        bump.reset();
+
+        inner_pool = Some(InnerStringPool::new(
+            bump,
+            |bump| RefCell::new(BumpVec::new_in(&bump)),
+        ));
+
+        swap(&mut self.0, &mut inner_pool);
+    }
+
+    pub(crate) fn store_c_string(
+        &self,
+        enc: &ENCODING,
+        buf: ExpatBufRef,
+    ) -> bool {
+        self.inner().rent(|vec| {
+            let mut vec = vec.borrow_mut();
+
+            if !self.rented_append(enc, buf, &mut *vec) {
+                return false;
+            }
+            if !self.rented_append_char('\0' as XML_Char, &mut *vec) {
+                return false;
+            }
+
+            true
+        })
+    }
+
+    pub(crate) fn append(
+        &self,
+        enc: &ENCODING,
+        mut read_buf: ExpatBufRef,
+    ) -> bool {
+        self.inner().rent(|vec| self.rented_append(enc, read_buf, &mut *vec.borrow_mut()))
+    }
+
+    fn rented_append(
+        &self,
+        enc: &ENCODING,
+        mut read_buf: ExpatBufRef,
+        vec: &mut BumpVec<XML_Char>,
+    ) -> bool {
+        let start = vec.as_ptr();
+
+        // REVIEW: Can this be replaced with self.is_empty() &&?
+        if start.is_null() && !self.rented_grow(vec) {
+            return false;
+        }
+
+        let mut cap = vec.capacity();
+        let mut len = vec.len();
+
+        loop {
+            debug_assert!(len <= cap);
+
+            // Continue to allocate if we don't have enough space
+            while (cap - len) < read_buf.len() {
+                if !self.rented_grow(vec) {
+                    return false;
+                }
+
+                cap = vec.capacity();
+            }
+
+            let start_len = vec.len();
+            let cap = vec.capacity();
+
+            vec.resize(cap, 0);
+
+            let mut write_buf = unsafe { ExpatBufRefMut::new_len(vec.as_mut_ptr().add(len), cap - start_len) };
+            let write_buf_len = write_buf.len();
+            let convert_res = unsafe {
+                XmlConvert!(enc, &mut read_buf, &mut write_buf)
+            };
+            // The write buf shrinks by how much was written to it
+            let written_size = write_buf_len - write_buf.len();
+
+            vec.truncate(start_len + written_size);
+
+            if convert_res == XML_Convert_Result::COMPLETED || convert_res == XML_Convert_Result::INPUT_INCOMPLETE {
+                break;
+            }
+        }
+
+        true
+    }
+
+    pub(crate) unsafe fn copy_c_string(
+        &self,
+        mut s: *const XML_Char,
+    ) -> Option<&mut [XML_Char]> {
+        // self.append_c_string(s);?
+        let successful = self.inner().rent(|vec| {
+            let mut vec = vec.borrow_mut();
+
+            loop {
+                if !self.rented_append_char(*s, &mut *vec) {
```

Comment:
As an idea, it would be great if we had a `CStringIterator` that turned a `*const XML_Char` into an `Iterator<XML_Char>`, and then this whole function could just be:
```Rust
vec.extend(CStringIterator::from(s));
```

Reply:
I also think `CStringIterator` could be useful in a bunch of other places in `xmlparse.rs`.

Reply:
> vec.extend(CStringIterator::from(s));

This does seem like a good idea, but I think that one problem here is that we need to call `try_reserve` first or else `extend` could panic due to OOM.

Reply:
And since this is a regular iterator with no knowledge of the length ahead of time, there's no way to get the length until after the iterator is used up

Reply:
Yeah, I guess we'd need something like `try_extend`, and that doesn't exist (I checked). We need to 1) find the length in advance, 2) call `try_reserve` on that many elements, and 3) then call `extend`.